File: DeckOfCards-Java/src/main/java/com/aprearo/deckofcards/Deck.java
package com.aprearo.deckofcards;

import java.util.ArrayList;

public class Deck {
    private ArrayList<Card> cards;

    public Deck() {
        cards = new ArrayList<>();
        for (Card.Suit suit : Card.Suit.values()) {
            for (Card.Rank rank : Card.Rank.values()) {
                cards.add(new Card(suit, rank));
            }
        }
    }

    public ArrayList<Card> getCards() {
        return cards;
    }


    public int size() {
        return  cards.size();
    }

    public void shuffle() {
        // Implementation of Fisher–Yates shuffle algorithm
        // http://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle

        int n = cards.size();
        for (int i = 0; i < n; i++) {
            int random = i + (int) (Math.random() * (n - i));
            Card randomCard = cards.get(random);
            cards.set(random, cards.get(i));
            cards.set(i, randomCard);
        }
    }

    public Card dealOne() {
        if (size() == 0) {
            throw new RuntimeException("All cards have been dealt");
        }
        Card card = cards.remove(0);
        return card;
    }

    public Card getCard(int index) {
        if (index < 0 || index >= cards.size()) {
            throw new RuntimeException(String.format("Card index not contained in deck: ", index));
        }
        return cards.get(index);
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        int size = size();
        for (int i = 0; i < size; i ++) {
            Card card = cards.get(i);
            builder.append(String.format("Card %d of %d: %s\n", i, size, card.toString()));
        }
        return builder.toString();
    }
}

File: DeckOfCards-Scala/src/main/scala/com/aprearo/deckofcards/Deck.scala
package com.aprearo.deckofcards

class Deck {
  var cards = List[Card]()

  Suit.values.foreach(suit =>
    Rank.values.foreach(rank =>
      cards = cards :+ new Card(suit, rank)
    )
  )

  def size: Int = { return cards.size }

  def shuffle = { cards = util.Random.shuffle(cards) }

  def dealOne: Card = {
    if (this.size == 0) {
      throw new RuntimeException("All cards have been dealt")
      return null
    }

    val card = cards.head
    cards = cards.tail
    return card
  }

  def getCard(index: Int): Card = {
    if ((index < 0) || (index > this.size - 1)) {
      throw new RuntimeException(s"Card index not contained in deck: ${index}")
      return null
    }

    return cards(index)
  }
}
